{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42981894-9575-4b1b-adc5-5bab26ea9801",
   "metadata": {
    "id": "42981894-9575-4b1b-adc5-5bab26ea9801"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/apps/miniconda3/envs/UnitedNet/lib/python3.9/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "2025-06-26 13:07:36,399\tINFO util.py:154 -- Missing packages: ['ipywidgets']. Run `pip install -U ipywidgets`, then restart the notebook server for rich notebook output.\n",
      "2025-06-26 13:07:36.871530: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2025-06-26 13:07:36.920508: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2025-06-26 13:07:36.935633: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2025-06-26 13:07:36.967803: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-06-26 13:07:38.760919: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import math\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import re\n",
    "\n",
    "sc.settings.verbosity = 0\n",
    "\n",
    "import GenKI as gk\n",
    "from GenKI.preprocesing import build_adata\n",
    "from GenKI.dataLoader import DataLoader\n",
    "from GenKI.train import VGAE_trainer\n",
    "from GenKI import utils\n",
    "\n",
    "import stringdb\n",
    "import networkx as nx\n",
    "\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy.sparse import issparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1afb1af9-3866-4d65-bd18-98e50c3ab0e7",
   "metadata": {
    "id": "1afb1af9-3866-4d65-bd18-98e50c3ab0e7"
   },
   "outputs": [],
   "source": [
    "adata = sc.read(\"VisiumRNABC_v3.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa56cd2d-d3aa-4855-92a6-5be2be5bcd91",
   "metadata": {
    "id": "fa56cd2d-d3aa-4855-92a6-5be2be5bcd91"
   },
   "outputs": [],
   "source": [
    "# # Load your deconvolution matrix\n",
    "# import pandas as pd\n",
    "\n",
    "# deconv_df = pd.read_excel(\"Cell_types_Probabilities.xlsx\", index_col=0)\n",
    "\n",
    "# # Optional: check alignment between adata.obs_names and deconv_df.index\n",
    "# matching_barcodes = adata.obs_names.intersection(deconv_df.index)\n",
    "# deconv_df = deconv_df.loc[matching_barcodes]\n",
    "\n",
    "# # Add the matrix to adata.obsm\n",
    "# adata.obsm[\"deconvolution\"] = deconv_df.values\n",
    "\n",
    "# # Store the column names (cell types)\n",
    "# adata.uns[\"deconvolution_celltypes\"] = deconv_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f50830f-34db-4949-9fc4-b6a9ed4c8032",
   "metadata": {
    "id": "1f50830f-34db-4949-9fc4-b6a9ed4c8032"
   },
   "outputs": [],
   "source": [
    "# adata.uns[\"deconvolution_celltypes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2096b0-ab78-4383-81de-e1ef709d6f98",
   "metadata": {
    "id": "0f2096b0-ab78-4383-81de-e1ef709d6f98"
   },
   "outputs": [],
   "source": [
    "# adata.obsm[\"deconvolution\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "816c8660-4d3c-4acd-b11e-e8edca9f1514",
   "metadata": {
    "id": "816c8660-4d3c-4acd-b11e-e8edca9f1514",
    "outputId": "643436d7-c482-4cde-9528-e76b3ba48f9c"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4011 × 500\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'spot_id', 'region', 'leiden'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'pca', 'spatial', 'spatialdata_attrs'\n",
       "    obsm: 'X_pca', 'spatial'\n",
       "    varm: 'PCs'\n",
       "    layers: 'raw_layer'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e80219b7-c0e7-45ea-b3e8-860625b36800",
   "metadata": {
    "id": "e80219b7-c0e7-45ea-b3e8-860625b36800"
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "# Standardize the data (zero mean, unit variance)\n",
    "sc.pp.scale(adata, zero_center=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cad1d69-a4cb-4010-b206-07cc9c1a90bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### Let's commence with creating a digital KO of Six2 expression in the DBiT-seq mouse embryo dataset.\n",
    "# gene_of_interest = \"TP53\"\n",
    "\n",
    "# # Verify that the gene of interest is part of the rownames in adata.var\n",
    "# if gene_of_interest in adata.var.index:\n",
    "#     print(f\"The gene {gene_of_interest} is present in the rownames of adata.var.\")\n",
    "# else:\n",
    "#     print(f\"The gene {gene_of_interest} is not present in the rownames of adata.var.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "934fae3f-42a9-47b5-a44f-c3a126d8a3f4",
   "metadata": {
    "id": "934fae3f-42a9-47b5-a44f-c3a126d8a3f4",
    "outputId": "e0ed332d-ae59-4f13-a319-8be65378103f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted adata.X to a sparse matrix.\n"
     ]
    }
   ],
   "source": [
    "# adata pre-processing to prepare for input in the GenKI tool\n",
    "adata.layers[\"norm\"] = adata.X.copy()\n",
    "\n",
    "# The adata.X should be normalised-scaled AND in sparse matrix format!\n",
    "if not issparse(adata.X):\n",
    "    sparse_matrix = csr_matrix(adata.X)\n",
    "    adata.X = sparse_matrix\n",
    "    print(\"Converted adata.X to a sparse matrix.\")\n",
    "else:\n",
    "    print(\"adata.X is already a sparse matrix.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a113ba53-c4ee-4b48-8677-a36a59523b03",
   "metadata": {
    "id": "a113ba53-c4ee-4b48-8677-a36a59523b03"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing gene: AREG\n",
      "use all the cells (4011) in adata\n",
      "loading GRN from \"GRNs/Visium500_breast_cancer_example.npz\"\n",
      "init completed\n",
      "\n",
      "Top 30 KO Responsive Genes for AREG:\n",
      "                dis  rank\n",
      "AREG      90.400504     1\n",
      "IL1RN      0.003106     2\n",
      "S100A8     0.003032     3\n",
      "CHI3L1     0.002981     4\n",
      "CPB1       0.002867     5\n",
      "S100A9     0.002743     6\n",
      "MMP12      0.002689     7\n",
      "SOCS3      0.002624     8\n",
      "GOLM1      0.002549     9\n",
      "IFIT2      0.002467    10\n",
      "MMP7       0.002395    11\n",
      "HERC5      0.002385    12\n",
      "TTC39A     0.002378    13\n",
      "SEMA3C     0.002317    14\n",
      "PARD6B     0.002253    15\n",
      "DHRS2      0.002244    16\n",
      "CA12       0.002189    17\n",
      "GABRP      0.002183    18\n",
      "CXCL11     0.002154    19\n",
      "AGR2       0.002127    20\n",
      "PTHLH      0.002097    21\n",
      "IGFBP3     0.002066    22\n",
      "REPS2      0.002037    23\n",
      "APOBEC3A   0.002023    24\n",
      "SYNPO2     0.002019    25\n",
      "CXCL14     0.001973    26\n",
      "HECTD2     0.001960    27\n",
      "FCRL2      0.001957    28\n",
      "SPP1       0.001948    29\n",
      "IGFBP5     0.001947    30\n",
      "\n",
      "Processing gene: BCL11A\n",
      "use all the cells (4011) in adata\n",
      "loading GRN from \"GRNs/Visium500_breast_cancer_example.npz\"\n",
      "init completed\n",
      "\n",
      "Top 30 KO Responsive Genes for BCL11A:\n",
      "               dis  rank\n",
      "BCL11A    1.524962     1\n",
      "KRT14     0.011763     2\n",
      "KRT17     0.011704     3\n",
      "ACTG2     0.009859     4\n",
      "CLU       0.009470     5\n",
      "MUCL1     0.009228     6\n",
      "REPS2     0.007848     7\n",
      "GPRC5A    0.007807     8\n",
      "AGR2      0.007129     9\n",
      "GFRA1     0.006248    10\n",
      "PGR       0.006126    11\n",
      "STC2      0.005976    12\n",
      "CRYAB     0.005788    13\n",
      "TACSTD2   0.005592    14\n",
      "KRT19     0.005513    15\n",
      "ACADSB    0.005437    16\n",
      "SFRP1     0.004980    17\n",
      "CXCL10    0.004903    18\n",
      "TFF3      0.004822    19\n",
      "CA12      0.004710    20\n",
      "ARHGAP29  0.004582    21\n",
      "NTN4      0.004515    22\n",
      "SFRP4     0.004502    23\n",
      "SDK2      0.004500    24\n",
      "KRT8      0.004494    25\n",
      "ITGBL1    0.004481    26\n",
      "ATP8B1    0.004453    27\n",
      "SEMA3C    0.004424    28\n",
      "PARD6B    0.004392    29\n",
      "ANKRD50   0.004271    30\n",
      "\n",
      "Processing gene: EPAS1\n",
      "use all the cells (4011) in adata\n",
      "loading GRN from \"GRNs/Visium500_breast_cancer_example.npz\"\n",
      "init completed\n",
      "\n",
      "Top 30 KO Responsive Genes for EPAS1:\n",
      "               dis  rank\n",
      "EPAS1     0.788955     1\n",
      "RGS5      0.003655     2\n",
      "KRT17     0.002835     3\n",
      "KRT14     0.002692     4\n",
      "MYLK      0.002669     5\n",
      "TPM2      0.002526     6\n",
      "TNC       0.002526     7\n",
      "ACTG2     0.002402     8\n",
      "ACTA2     0.002299     9\n",
      "CRYAB     0.001986    10\n",
      "ADAMTS1   0.001862    11\n",
      "SFRP1     0.001839    12\n",
      "MCAM      0.001794    13\n",
      "TAGLN     0.001522    14\n",
      "MYL9      0.001479    15\n",
      "AQP1      0.001399    16\n",
      "EGFL7     0.001071    17\n",
      "SPARCL1   0.001049    18\n",
      "HTRA1     0.001040    19\n",
      "EHD2      0.000994    20\n",
      "ANKRD30A  0.000886    21\n",
      "PDGFRB    0.000871    22\n",
      "FSTL1     0.000840    23\n",
      "ENG       0.000774    24\n",
      "LIMS2     0.000707    25\n",
      "FABP4     0.000701    26\n",
      "CCL14     0.000549    27\n",
      "ACKR1     0.000518    28\n",
      "C7        0.000381    29\n",
      "KLF2      0.000345    30\n",
      "\n",
      "Results have been saved to Top_RNA_Niche_Protein_Interactions_UnitedNet.xlsx\n",
      "\n",
      "Results have been saved to Top_RNA_Niche_Protein_Interactions_UnitedNet_StringDB_filtered.xlsx\n",
      "Genes of interest: ['AREG', 'BCL11A', 'EPAS1']\n",
      "\n",
      "Processing gene of interest: AREG\n",
      "Targets from Excel for AREG: 130 genes\n",
      "Loaded 30 responsive genes from Top30_Responsive_Genes/Top30_Responsive_Genes_AREG.csv.\n",
      "Total combined genes before removing duplicates: 160.\n",
      "Total combined genes after removing duplicates: 104.\n",
      "Combined list saved to Combined_Genes_List_AREG.txt.\n",
      "\n",
      "Processing gene of interest: BCL11A\n",
      "Targets from Excel for BCL11A: 140 genes\n",
      "Loaded 30 responsive genes from Top30_Responsive_Genes/Top30_Responsive_Genes_BCL11A.csv.\n",
      "Total combined genes before removing duplicates: 170.\n",
      "Total combined genes after removing duplicates: 106.\n",
      "Combined list saved to Combined_Genes_List_BCL11A.txt.\n",
      "\n",
      "Processing gene of interest: EPAS1\n",
      "Targets from Excel for EPAS1: 80 genes\n",
      "Loaded 30 responsive genes from Top30_Responsive_Genes/Top30_Responsive_Genes_EPAS1.csv.\n",
      "Total combined genes before removing duplicates: 110.\n",
      "Total combined genes after removing duplicates: 78.\n",
      "Combined list saved to Combined_Genes_List_EPAS1.txt.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "combined_results = []\n",
    "#all_genes_of_interest = ['PGR', 'TACSTD2', 'KRT14','KRT17','CXCL9','CXCL10','CXCL13']\n",
    "all_genes_of_interest = ['AREG', 'BCL11A', 'EPAS1']\n",
    "num_GenKI_KO_responsive_genes = 30\n",
    "num_shap = 5 \n",
    "\n",
    "\n",
    "for gene_of_interest in all_genes_of_interest:\n",
    "    print(f\"\\nProcessing gene: {gene_of_interest}\")\n",
    "\n",
    "    data_wrapper =  DataLoader(\n",
    "                adata, # adata object\n",
    "                target_gene = [gene_of_interest], # KO gene name\n",
    "                target_cell = None, # obsname for cell type, if none use all\n",
    "                obs_label = \"ident\", # colname for genes\n",
    "                GRN_file_dir = \"GRNs\", # folder name for GRNs\n",
    "                rebuild_GRN = False, # whether build GRN by pcNet\n",
    "                pcNet_name = \"Visium500_breast_cancer_example\", # GRN file name\n",
    "                verbose = True, # whether verbose\n",
    "                n_cpus = 20, # multiprocessing\n",
    "                )\n",
    "\n",
    "    data_wt = data_wrapper.load_data()\n",
    "    data_ko = data_wrapper.load_kodata()\n",
    "\n",
    "\n",
    "    hyperparams = {\n",
    "        \"epochs\": 100,  # Increased epochs for more training\n",
    "        \"lr\": 7e-4,  # Adjusted learning rate\n",
    "        \"beta\": 1e-4,  # Increased beta for stronger regularization\n",
    "        \"seed\": 8096  # Trying a different seed\n",
    "    }\n",
    "\n",
    "\n",
    "    log_dir = None\n",
    "\n",
    "    sensei = VGAE_trainer(\n",
    "        data_wt,\n",
    "        epochs=hyperparams[\"epochs\"],\n",
    "        lr=hyperparams[\"lr\"],\n",
    "        log_dir=log_dir,\n",
    "        beta=hyperparams[\"beta\"],\n",
    "        seed=hyperparams[\"seed\"],\n",
    "        verbose=False,\n",
    "    )\n",
    "\n",
    "    sensei.train()\n",
    "    #sensei.save_model('Visium500_breast_cancer_model_example')\n",
    "\n",
    "    # Get latent variables for KO data\n",
    "    z_mu_ko, z_std_ko = sensei.get_latent_vars(data_ko)\n",
    "    z_mu_wt, z_std_wt = sensei.get_latent_vars(data_wt)\n",
    "\n",
    "    # Calculate the distance between WT and KO data\n",
    "    dis = gk.utils.get_distance(z_mu_ko, z_std_ko, z_mu_wt, z_std_wt, by=\"KL\")\n",
    "\n",
    "    # Get the ranked list of responsive genes\n",
    "    res_raw = utils.get_generank(data_wt, dis, rank=True)\n",
    "\n",
    "    # Store the top 30 responsive genes\n",
    "    top_genes = res_raw.head(num_GenKI_KO_responsive_genes)\n",
    "    print(f\"Top {num_GenKI_KO_responsive_genes} KO Responsive Genes for {gene_of_interest}:\\n{top_genes}\")\n",
    "\n",
    "    os.makedirs('Top_Responsive_Genes', exist_ok=True)\n",
    "    top_genes_path = f'Top{num_GenKI_KO_responsive_genes}_Responsive_Genes/Top{num_GenKI_KO_responsive_genes}_Responsive_Genes_{gene_of_interest}.csv'\n",
    "    top_genes.to_csv(top_genes_path)\n",
    "\n",
    "    # Extract the top genes for the current KO gene\n",
    "    genki_list = top_genes.index.tolist()\n",
    "\n",
    "    # Load the data from CSV file\n",
    "    file_path = 'feature_feature_importance.csv'  # Replace with your file path\n",
    "    data = pd.read_csv(file_path)\n",
    "\n",
    "    # Filter data for the genes of interest\n",
    "    filtered_data = data[data['Source'].isin(genki_list)]\n",
    "\n",
    "    # Separate data for RNA -> Niche and RNA -> Protein directions\n",
    "    rna_niche_data = filtered_data[filtered_data['Direction'] == 'RNA -> Niche']\n",
    "    rna_protein_data = filtered_data[filtered_data['Direction'] == 'RNA -> Protein']\n",
    "\n",
    "    # Find the top unique RNA -> Niche and RNA -> Protein interactions with the highest 'Value'\n",
    "    top_rna_niche = (\n",
    "        rna_niche_data\n",
    "        .sort_values(by='Value', ascending=False)\n",
    "        .drop_duplicates(subset=['Source', 'Target'])\n",
    "        .groupby('Source')\n",
    "        .head(num_shap)\n",
    "    )\n",
    "\n",
    "    top_rna_protein = (\n",
    "        rna_protein_data\n",
    "        .sort_values(by='Value', ascending=False)\n",
    "        .drop_duplicates(subset=['Source', 'Target'])\n",
    "        .groupby('Source')\n",
    "        .head(num_shap)\n",
    "    )\n",
    "\n",
    "    # Identify cases where the 'Target' value is the same as the 'Source' value\n",
    "    rna_protein_duplicates = top_rna_protein[top_rna_protein['Target'] == top_rna_protein['Source']]\n",
    "    rna_niche_duplicates = top_rna_niche[top_rna_niche['Target'] == top_rna_niche['Source']]\n",
    "\n",
    "    # Remove duplicate entries from the original top lists\n",
    "    top_rna_protein_cleaned = top_rna_protein[~(top_rna_protein['Target'] == top_rna_protein['Source'])]\n",
    "    top_rna_niche_cleaned = top_rna_niche[~(top_rna_niche['Target'] == top_rna_niche['Source'])]\n",
    "\n",
    "    # Find the next highest entries for those with matching 'Target' and 'Source'\n",
    "    next_rna_protein = (\n",
    "        rna_protein_data\n",
    "        .loc[~rna_protein_data.index.isin(rna_protein_duplicates.index)]\n",
    "        .sort_values(by='Value', ascending=False)\n",
    "    )\n",
    "\n",
    "    next_rna_protein_add = (\n",
    "        next_rna_protein\n",
    "        .groupby('Source')\n",
    "        .apply(lambda x: x[~x['Target'].isin(top_rna_protein_cleaned['Target'])].head(1))\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    next_rna_niche = (\n",
    "        rna_niche_data\n",
    "        .loc[~rna_niche_data.index.isin(rna_niche_duplicates.index)]\n",
    "        .sort_values(by='Value', ascending=False)\n",
    "    )\n",
    "\n",
    "    next_rna_niche_add = (\n",
    "        next_rna_niche\n",
    "        .groupby('Source')\n",
    "        .apply(lambda x: x[~x['Target'].isin(top_rna_niche_cleaned['Target'])].head(1))\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    # Combine the original top lists with the added entries\n",
    "    final_top_rna_protein = (\n",
    "        pd.concat([top_rna_protein_cleaned, next_rna_protein_add])\n",
    "        .sort_values(by=['Source', 'Value'], ascending=[True, False])\n",
    "        .groupby('Source')\n",
    "        .head(num_shap)\n",
    "    )\n",
    "\n",
    "    final_top_rna_niche = (\n",
    "        pd.concat([top_rna_niche_cleaned, next_rna_niche_add])\n",
    "        .sort_values(by=['Source', 'Value'], ascending=[True, False])\n",
    "        .groupby('Source')\n",
    "        .head(num_shap)\n",
    "    )\n",
    "\n",
    "    # Combine all results into a single DataFrame\n",
    "    combined_df = pd.concat([final_top_rna_protein, final_top_rna_niche])\n",
    "\n",
    "    # Add a column to indicate the KO gene\n",
    "    combined_df['KO_Gene'] = gene_of_interest\n",
    "\n",
    "    # Append to the list\n",
    "    combined_results.append(combined_df)\n",
    "\n",
    "# Concatenate all results\n",
    "all_combined_df = pd.concat(combined_results)\n",
    "\n",
    "# Save the combined results to a single sheet in an Excel file\n",
    "output_path = 'Top_RNA_Niche_Protein_Interactions_UnitedNet.xlsx'\n",
    "\n",
    "with pd.ExcelWriter(output_path) as writer:\n",
    "    all_combined_df.to_excel(writer, sheet_name='Top Interactions', index=False)\n",
    "\n",
    "print(f\"\\nResults have been saved to {output_path}\")\n",
    "\n",
    "## Rest Process\n",
    "\n",
    "\n",
    "interactions_df = pd.read_excel('Top_RNA_Niche_Protein_Interactions_UnitedNet.xlsx', sheet_name='Top Interactions')\n",
    "\n",
    "# Concatenate both columns and drop duplicates\n",
    "genes = pd.unique(pd.concat([interactions_df[\"Target\"], interactions_df[\"Source\"]]))\n",
    "\n",
    "# If you want a Python list instead of a NumPy array\n",
    "genes = genes.tolist()\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import stringdb\n",
    "\n",
    "species = 9606\n",
    "\n",
    "# Step 1: Get STRING IDs\n",
    "string_ids_df = stringdb.get_string_ids(genes)\n",
    "string_ids = string_ids_df.queryItem.tolist()\n",
    "\n",
    "# Step 2: Get interaction network\n",
    "def get_interactions(string_ids, species):\n",
    "    url = \"https://string-db.org/api/json/network\"\n",
    "    params = {\n",
    "        \"identifiers\": \"%0d\".join(string_ids),  # line-break separated\n",
    "        \"species\": species\n",
    "    }\n",
    "    r = requests.get(url, params=params)\n",
    "    return pd.DataFrame(r.json())\n",
    "\n",
    "# Step 3: Format output\n",
    "interactions_string_df = get_interactions(string_ids, species)\n",
    "df_edges = interactions_string_df[[\"preferredName_A\", \"preferredName_B\", \"score\"]]\n",
    "df_edges.columns = [\"source\", \"target\", \"combined_score\"]\n",
    "\n",
    "# Rename for clarity if needed\n",
    "df_edges = df_edges.rename(columns={\n",
    "    \"source\": \"Source\",\n",
    "    \"target\": \"Target\"\n",
    "})\n",
    "\n",
    "# Merge: Left join from your original df (keep all rows from df, add combined_score if available)\n",
    "df_merged = interactions_df.merge(df_edges, on=[\"Target\", \"Source\"], how=\"left\")\n",
    "\n",
    "#interaction_df = df_merged.dropna(subset=['combined_score'])\n",
    "\n",
    "# Save the combined results to a single sheet in an Excel file\n",
    "output_path = 'Top_RNA_Niche_Protein_Interactions_UnitedNet_StringDB_filtered.xlsx'\n",
    "\n",
    "with pd.ExcelWriter(output_path) as writer:\n",
    "    all_combined_df.to_excel(writer, sheet_name='Top Interactions', index=False)\n",
    "\n",
    "print(f\"\\nResults have been saved to {output_path}\")\n",
    "\n",
    "\n",
    "interactions_df = pd.read_excel('Top_RNA_Niche_Protein_Interactions_UnitedNet_StringDB_filtered.xlsx', sheet_name='Top Interactions')\n",
    "targets = interactions_df['Target'].tolist()\n",
    "\n",
    "all_genes_of_interest = ['AREG', 'BCL11A', 'EPAS1']\n",
    "print(f\"Genes of interest: {all_genes_of_interest}\")\n",
    "\n",
    "# Build targets list per KO gene as a dictionary\n",
    "targets_dict = {\n",
    "    ko_gene: interactions_df[interactions_df['KO_Gene'] == ko_gene]['Target'].tolist()\n",
    "    for ko_gene in all_genes_of_interest\n",
    "}\n",
    "\n",
    "# Process each gene of interest\n",
    "for gene_of_interest in all_genes_of_interest:\n",
    "    print(f\"\\nProcessing gene of interest: {gene_of_interest}\")\n",
    "\n",
    "    # Get the targets for this KO gene\n",
    "    targets = targets_dict[gene_of_interest]\n",
    "    print(f\"Targets from Excel for {gene_of_interest}: {len(targets)} genes\")\n",
    "\n",
    "    # Initialize list of responsive genes from Top_Responsive_Genes\n",
    "    responsive_genes = []\n",
    "\n",
    "    file_name = f'Top{num_GenKI_KO_responsive_genes}_Responsive_Genes/Top{num_GenKI_KO_responsive_genes}_Responsive_Genes_{gene_of_interest}.csv'\n",
    "    if os.path.exists(file_name):\n",
    "        df_top_genes = pd.read_csv(file_name, index_col=0)\n",
    "        responsive_genes = df_top_genes.index.tolist()\n",
    "        print(f\"Loaded {len(responsive_genes)} responsive genes from {file_name}.\")\n",
    "    else:\n",
    "        print(f\"File {file_name} does not exist. Skipping Top responsive genes.\")\n",
    "\n",
    "    # Combine targets + responsive_genes\n",
    "    combined_list = targets + responsive_genes\n",
    "    print(f\"Total combined genes before removing duplicates: {len(combined_list)}.\")\n",
    "\n",
    "    # Remove duplicates\n",
    "    combined_list = list(set(combined_list))\n",
    "    print(f\"Total combined genes after removing duplicates: {len(combined_list)}.\")\n",
    "\n",
    "    # Save to a file\n",
    "    output_file = f'Combined_Genes_List_{gene_of_interest}.txt'\n",
    "    with open(output_file, 'w') as f:\n",
    "        for gene in combined_list:\n",
    "            f.write(f\"{gene}\\n\")\n",
    "\n",
    "    print(f\"Combined list saved to {output_file}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "815f0e63-74c8-4268-bd73-f9e94f678882",
   "metadata": {
    "id": "815f0e63-74c8-4268-bd73-f9e94f678882"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "OJTY-Les3pIa",
   "metadata": {
    "id": "OJTY-Les3pIa"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9IniRcCD3pGS",
   "metadata": {
    "id": "9IniRcCD3pGS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "IN-AwXce3pEO",
   "metadata": {
    "id": "IN-AwXce3pEO"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eLZfBALr3pB7",
   "metadata": {
    "id": "eLZfBALr3pB7"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "752w9VrJ3o_t",
   "metadata": {
    "id": "752w9VrJ3o_t"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pDVehT_y3o9J",
   "metadata": {
    "id": "pDVehT_y3o9J"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dW8h44xK3o6z",
   "metadata": {
    "id": "dW8h44xK3o6z"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "RvItFUqD3o4f",
   "metadata": {
    "id": "RvItFUqD3o4f"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "811bc780-42bf-4a2c-831e-b97a8782ff0e",
   "metadata": {
    "id": "811bc780-42bf-4a2c-831e-b97a8782ff0e"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40379a9e-0fb1-4371-a3e3-b54097a9bf34",
   "metadata": {
    "id": "40379a9e-0fb1-4371-a3e3-b54097a9bf34"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdf23fb0-0d0f-43d1-a72b-046c6f1eb2f0",
   "metadata": {
    "id": "fdf23fb0-0d0f-43d1-a72b-046c6f1eb2f0"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d46f6e0-c6b2-4ded-a216-d455943d172e",
   "metadata": {
    "id": "4d46f6e0-c6b2-4ded-a216-d455943d172e"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e65009-d07a-4448-9680-7f21d792eb9c",
   "metadata": {
    "id": "10e65009-d07a-4448-9680-7f21d792eb9c"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fda25d74-dea6-4621-8b46-693c6be53eb0",
   "metadata": {
    "id": "fda25d74-dea6-4621-8b46-693c6be53eb0"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5a04566-e923-4a70-bf04-b8a578c26a32",
   "metadata": {
    "id": "a5a04566-e923-4a70-bf04-b8a578c26a32"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c62a39d8-f406-441c-8f13-4cfbb7b627e0",
   "metadata": {
    "id": "c62a39d8-f406-441c-8f13-4cfbb7b627e0"
   },
   "outputs": [],
   "source": [
    "# Let's create a STRINGdb network of the targets affected by the Six2 digital KO on the DBiT-seq dataset\n",
    "genes = list(res_raw.index[1:10,])\n",
    "\n",
    "string_ids = stringdb.get_string_ids(genes, species = 9606)   # human is 'species=9606'\n",
    "\n",
    "# Performe pathway enrichment\n",
    "enrichment_df = stringdb.get_enrichment(string_ids.queryItem, species = 9606)\n",
    "enrichment_df\n",
    "df = enrichment_df\n",
    "sorted_df = df.sort_values(by='fdr', ascending=False)\n",
    "\n",
    "# Select the top 10 pathways to visualise\n",
    "top_10_pathways = sorted_df['description'].head(10)\n",
    "\n",
    "print(\"Top 10 Pathways:\")\n",
    "print(top_10_pathways)\n",
    "\n",
    "# Plotting negative log10 of FDR values against description to discern most significant pathways\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.barh(sorted_df['description'].head(10), -1 * sorted_df['fdr'].head(10).apply(lambda x: math.log10(x)))\n",
    "plt.xlabel('-log10(FDR)')\n",
    "plt.ylabel('Description')\n",
    "plt.title('Top 10 Pathways: Negative Log10 FDR values by Description')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f05262-d74e-4602-8713-1844d6d30200",
   "metadata": {
    "id": "f9f05262-d74e-4602-8713-1844d6d30200"
   },
   "outputs": [],
   "source": [
    "# Preparing visualization of STRINGdb network with Networkx\n",
    "network_df = stringdb.get_network(string_ids.queryItem, species = 9606, required_score=400)\n",
    "df = network_df\n",
    "G = nx.Graph()\n",
    "for index, row in df.iterrows():\n",
    "    G.add_edge(row['preferredName_A'], row['preferredName_B'], weight=row['score'])\n",
    "\n",
    "# Draw the graph\n",
    "pos = nx.spring_layout(G)  # positions for all nodes\n",
    "nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=10)\n",
    "\n",
    "# Draw edge labels with weights\n",
    "edge_labels = nx.get_edge_attributes(G, 'weight')\n",
    "nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dddc193-b8da-4232-bac0-1bacd054c103",
   "metadata": {
    "id": "8dddc193-b8da-4232-bac0-1bacd054c103"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ee020fe-aa61-409b-ae19-5e6898efaea7",
   "metadata": {
    "id": "7ee020fe-aa61-409b-ae19-5e6898efaea7"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abf3ddc6-6c53-45f2-b2c7-7d9c5be5bea1",
   "metadata": {
    "id": "abf3ddc6-6c53-45f2-b2c7-7d9c5be5bea1"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b256008f-9bc0-49d7-8655-feb4e3084846",
   "metadata": {
    "id": "b256008f-9bc0-49d7-8655-feb4e3084846"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b01f0587-58a5-4b38-8adf-035a792d4ea9",
   "metadata": {
    "id": "b01f0587-58a5-4b38-8adf-035a792d4ea9"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d61772-d8f5-4c2e-b2de-c917b0314e0f",
   "metadata": {
    "id": "14d61772-d8f5-4c2e-b2de-c917b0314e0f"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bff7a55-8db4-4160-91e5-9671db6ec167",
   "metadata": {
    "id": "4bff7a55-8db4-4160-91e5-9671db6ec167"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f0e14ff-0942-4243-a50c-9cf21fe1501d",
   "metadata": {
    "id": "4f0e14ff-0942-4243-a50c-9cf21fe1501d"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc8f4a20-c554-4230-bb21-91ed246e5cb2",
   "metadata": {
    "id": "bc8f4a20-c554-4230-bb21-91ed246e5cb2"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59e4a287-3799-4a13-9b36-2436513e759d",
   "metadata": {
    "id": "59e4a287-3799-4a13-9b36-2436513e759d"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fc193ffc-652d-43e4-b83d-525f69f5c3f7",
   "metadata": {
    "id": "fc193ffc-652d-43e4-b83d-525f69f5c3f7"
   },
   "source": [
    "#"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
